### PR TITLE
Backtesting updated to provide a more accurate fill prices

### PR DESCRIPTION
**Pull request by davidlatte**
Applies to  for market, limit, and stop loss orders.

### CODE REVIEW

File: lumibot/backtesting/backtesting_broker.py [hunk 1]
```diff
@@ -557,39 +550,39 @@ def process_pending_orders(self, strategy):
...
+        # Gap Down case: Limit wasn't triggered by previous candle but current candle opens lower, fill it now
+        if side == "buy" and limit_price >= open_:
+            return open_
+
+        # Current candle triggered limit normally
+        if low <= limit_price <= high:
+            return limit_price
+
+        # Limit has not been met
+        return 0.0
```
Comment (grzesir) [inline lumibot/backtesting/backtesting_broker.py:568]: should return None, not 0. 0 can actually be a fill price is some extreme situations

File: lumibot/backtesting/backtesting_broker.py [hunk 2]
```diff
@@ -557,39 +550,39 @@ def process_pending_orders(self, strategy):
...
+        # Gap Up case: Stop wasn't triggered by previous candle but current candle opens higher, fill it now
+        if side == "buy" and stop_price <= open_:
+            return open_
+
+        # Current candle triggered stop normally
+        if low <= stop_price <= high:
+            return stop_price
+
+        # Stop has not been met
+        return 0.0
```
Comment (grzesir) [inline lumibot/backtesting/backtesting_broker.py:585]: should return None, not 0. 0 can actually be a fill price is some extreme situations

File: lumibot/backtesting/backtesting_broker.py [hunk 3]
```diff
@@ -468,55 +468,48 @@ def process_pending_orders(self, strategy):
...
+                price = self.stop_order(order.stop_price, order.side, open, high, low)
+
             elif order.type == "stop_limit":
                 if not order.price_triggered:
                     price = self.stop_order(
-                        order.stop_price, order.side, close, high, low
+                        order.stop_price, order.side, open, high, low
                     )
-                    if price != None:
+                    if price:
```
Comment (grzesir) [inline lumibot/backtesting/backtesting_broker.py:492]: wouldn't this also filter out $0 fills?

Comment (davidlatte) [reply lumibot/backtesting/backtesting_broker.py:492]: Technically yes ... but NYSE/NASDAQ/etc don't allow any $0 prices, right?

File: lumibot/backtesting/backtesting_broker.py [hunk 4]
```diff
@@ -468,55 +468,48 @@ def process_pending_orders(self, strategy):
...
-                            
-                    # Check if we have hit the trail stop price for a buy order
-                    elif order.side == "buy":
-                        if high >= order._trail_stop_price:
-                            price = high
-                
-                # Get the assets last price
-                asset_price = self.get_last_price(order.asset)
-                
+                if order._trail_stop_price:
```
Comment (grzesir) [inline lumibot/backtesting/backtesting_broker.py:503]: issue with None?

File: lumibot/backtesting/backtesting_broker.py [hunk 5]
```diff
@@ -525,9 +518,9 @@ def process_pending_orders(self, strategy):
             #############################
             # Fill the order.
             #############################
-            
-            # If the price is not None, then the order has been filled
-            if price is not None:
+
+            # If the price is set, then the order has been filled
+            if price:
```
Comment (grzesir) [inline lumibot/backtesting/backtesting_broker.py:523]: Issue with $0/None?

File: lumibot/backtesting/backtesting_broker.py (hunk 2 above)
Comment (davidlatte) [reply lumibot/backtesting/backtesting_broker.py]: Reverted back to using None

File: lumibot/backtesting/backtesting_broker.py (hunk 3 above)
Comment (davidlatte) [reply lumibot/backtesting/backtesting_broker.py]: Reverted back to using None

**Pull request merged**